### PR TITLE
✨ DD-based Simulator QDMI device

### PR DESCRIPTION
**Pull request by burgholzer**
## Description

This PR adds a new QDMI device to the available devices in MQT Core:
A classical simulator device based on the same Decision Diagram (DD) approach that also powers MQT DDSIM.
The device supports OpenQASM2 and OpenQASM3 as well as a wide variety of gates.
It supports weak simulation (=sampling) and strong simulation (=state vector simulation) based on the specified number of shots (`0` shots indicate that strong simulation is desired).
The sampling mode fully supports dynamic quantum circuits with mid-circuit measurements, resets, and conditional gates.

This marks the first MQT QDMI device that implements the full job interface.

At the moment this is still an early draft that just provides the first draft of the corresponding device. It lacks proper embedding as well as testing within MQT Core.

## Checklist:

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [x] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

### CODE REVIEW

File: test/qdmi/dd/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  generate_device_defs_executable(${QDMI_PREFIX})
+  # Set test target name
+  set(TARGET_NAME ${MQT_CORE_TARGET_NAME}-qdmi-ddsim-device-test)
+  # Add the test executable
+  package_add_test(${TARGET_NAME} MQT::CoreQDMI_DDSIM_Device test_device.cpp)
+  # Add dependency to enforce implementation of all interface functions
+  add_dependencies(${TARGET_NAME} qdmi_test_mqt_ddsim_device_defs)
+  # Set the executable path and version information
+  target_compile_definitions(${TARGET_NAME} PRIVATE MQT_CORE_VERSION="${MQT_CORE_VERSION}")
+endif()
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply test/qdmi/dd/CMakeLists.txt:35]: This is a false positive. The respective function is defined as part of the QDMI library which is being pulled in via FetchContent. Specifically in https://github.com/Munich-Quantum-Software-Stack/QDMI/blob/develop/cmake/PrefixHandling.cmake

File: test/qdmi/dd/helpers/test_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  }
+  auto keyVec = splitCSV(keys);
+  const size_t vs = querySize(job, QDMI_JOB_RESULT_HIST_VALUES);
+  std::vector<size_t> vals(vs / sizeof(size_t));
+  if (vs > 0) {
+    [[maybe_unused]] int rc = MQT_DDSIM_QDMI_device_job_get_results(
+        job, QDMI_JOB_RESULT_HIST_VALUES, vs, vals.data(), nullptr);
+    assert(rc == QDMI_SUCCESS);
+  }
+  return {std::move(keyVec), std::move(vals)};
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply test/qdmi/dd/helpers/test_utils.cpp:161]: std::string has an implicit `\0` terminator. QDMI returns the buffer size that includes the zero terminator. So `ks-1` is correct in the string allocation, as it will implicitly contain one more byte that is already initialised to `\0`.

[the remark this answers is not in the document]
Comment (burgholzer) [reply test/qdmi/dd/helpers/test_utils.cpp:161]: The last byte being written in the data buffer is always going to be a zero terminator `\0`, which is what is already at position `ks-1`. The data reported by QDMI is a string that would report size `ks-1`.

File: test/qdmi/dd/error_handling_test.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  EXPECT_EQ(MQT_DDSIM_QDMI_device_job_get_results(
+                j.job, QDMI_JOB_RESULT_CUSTOM3, 0, nullptr, nullptr),
+            QDMI_ERROR_NOTSUPPORTED);
+  EXPECT_EQ(MQT_DDSIM_QDMI_device_job_get_results(
+                j.job, QDMI_JOB_RESULT_CUSTOM4, 0, nullptr, nullptr),
+            QDMI_ERROR_NOTSUPPORTED);
+  EXPECT_EQ(MQT_DDSIM_QDMI_device_job_get_results(
+                j.job, QDMI_JOB_RESULT_CUSTOM5, 0, nullptr, nullptr),
+            QDMI_ERROR_NOTSUPPORTED);
+}
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply test/qdmi/dd/error_handling_test.cpp:251]: Parameters on sessions may only be set _before_ initialisation. Once `init` has been called on the session, any call to set a parameter for the session would trigger `QDMI_ERROR_BADSTATE`. And since `SessionGuard` directly initialises the session, a separate session is needed for testing these parameter corner cases.

File: test/qdmi/dd/error_handling_test.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+  ASSERT_EQ(qdmi_test::setProgram(j.job, QDMI_PROGRAM_FORMAT_QASM3,
+                                  qdmi_test::QASM3_Bell_Sampling),
+            QDMI_SUCCESS);
+  ASSERT_EQ(qdmi_test::setShots(j.job, 16), QDMI_SUCCESS);
+  ASSERT_EQ(qdmi_test::submitAndWait(j.job, 0), QDMI_SUCCESS);
+  EXPECT_EQ(MQT_DDSIM_QDMI_device_job_get_results(j.job, QDMI_JOB_RESULT_MAX, 0,
+                                                  nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT);
+}
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply test/qdmi/dd/error_handling_test.cpp:127]: Similar reasoning as for the Custom parameters also applies here. Parameters may only be set _before_ initialization.

File: src/qdmi/dd/Device.cpp [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+        const auto nqubits = qc.getNqubits();
+        dd_ = std::make_unique<dd::Package>(nqubits);
+        stateVecDD_ = dd::simulate(qc, dd::makeZeroState(nqubits, *dd_), *dd_);
+        status_.store(QDMI_JOB_STATUS_DONE);
+      } catch (const std::exception& e) {
+        status_.store(QDMI_JOB_STATUS_FAILED);
+        std::cerr << "Error: " << e.what() << '\n';
+      }
+      qdmi::dd::Device::get().decreaseRunningJobs();
+    });
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply src/qdmi/dd/Device.cpp:521]: The OpenQASM parser in MQT Core is backwards compatible with OpenQASM 2. Hence, there is no regression here and the importer works for both program versions.
